### PR TITLE
Wikitext - [WIP]

### PR DESCRIPTION
**Pull request by ayushi-3536**

### CODE REVIEW

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+
+__version__ = '0.0.1'
+
+logger = logging.getLogger('LM_Bench')
+
+
+class LanguageModelBenchmark(AbstractMultiObjectiveBenchmark):
+
+    def __init__(self, rng: Union[np.random.RandomState, int, None] = None, **kwargs):
+        super(LanguageModelBenchmark, self).__init__(rng=rng)
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:41]: Can you please add a short description?

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:41]: Can you please add a short description?

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+        self.__seed_everything()
+        self.variable = {"eval_batch_size": 10,
+                         "nlayers": 2,
+                         "bptt": 35,
+                         "tied": True,
+                         # number of attention head
+                         "nhead": 2,
+                         "ntoken": self.ntokens
+                         }
+        print("len of corpus dict", self.ntokens)
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: Please use logging.debug for printing

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: Please use logging.debug for printing

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+
+    def __seed_everything(self):
+        """Helperfunction: Make the benchmark deterministic by setting the correct seeds"""
+        seed = self.rng.randint(0, 100000)
+        print("seed obtained", seed)
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: same here

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: same here

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+            ),
+            CS.UniformFloatHyperparameter(
+                'clip', default_value=0.99, lower=0.1, upper=2
+            )
+
+        ])
+        return cs
+
+    @staticmethod
+    def get_objective_names(self) -> List[str]:
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: Remove `self`.

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: Remove `self`.

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+        return ['log_perplexity', 'accuracy', 'time']
+
+    @staticmethod
+    def get_fidelity_space(seed: Union[int, None] = None) -> CS.ConfigurationSpace:
+
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:116]: Short docstring please. (numpy format: sorry i have mixed up `google`with `numpy`)

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:116]: Short docstring please. (numpy format: sorry i have mixed up `google`with `numpy`)

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+                           'journal={arXiv preprint arXiv:2106.12639},'
+                           'year={2021}',
+                           ],
+        }
+
+    def init_model(self, config: Union[CS.Configuration, Dict]):
+        """ Function that returns the model initialized based on the configuration and fidelity
+        """
+
+        if isinstance(config, CS.Configuration):
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: actually we don't need that check (`if isinstance`) here. 

The objective_function -wrapper casts configurations always to dicts.

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: actually we don't need that check (`if isinstance`) here. 

The objective_function -wrapper casts configurations always to dicts.

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+                Note: the number of epoch is 1 indexed! (Results after the first epoch: epoch = 1)
+
+            Fidelity parameters, check get_fidelity_space(). Uses default (max) value if None.
+        rng : np.random.RandomState, int, None
+            Random seed to use in the benchmark.
+
+            To prevent overfitting on a single seed, it is possible to pass a
+            parameter ``rng`` as 'int' or 'np.random.RandomState' to this function.
+            If this parameter is not given, the default random state is used.
+
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:170]: Add `shuffle` here.

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:170]: Add `shuffle` here.

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 8]
```diff
@@ -0,0 +1,368 @@
...
+        t = tqdm.tqdm(total=epochs)
+        for epoch in range(epochs):
+            epoch_start_time = time.time()
+            train_loss, train_acc = model.train_fun(self.ntokens, criterion, train_data, learning_rate, clip)
+            train_time += time.time() - epoch_start_time
+            start = time.time()
+            val_loss, val_acc = model.eval_fun(self.ntokens, criterion, val_data)
+            val_loss = np.clip(val_loss, 1e-10, 10)
+            print("val acc for last epoch", val_acc)
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: `logging.debug`
see above

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: `logging.debug`
see above

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,368 @@
...
+            eval_time += start - time.time()
+
+            t.set_postfix(val_accuracy=val_acc)
+            t.update()
+
+            if not np.isfinite(val_loss):
+                val_loss = 7
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:233]: is there a reason for the 7? 

Could you please add a comment to it, if that is from the authors

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:233]: is there a reason for the 7? 

Could you please add a comment to it, if that is from the authors

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,368 @@
...
+                learning_rate /= learning_rate_factor
+
+        start_time = time.time()
+        _, test_acc = model.eval_fun(self.ntokens, criterion, test_data)
+        eval_test_runtime = time.time() - start_time
+
+        perplexity = math.exp(best_val_loss)
+        log_perplexity = best_val_loss
+        neg_log_perplexity = 10 - best_val_loss
+        elapsed_time = float(ts_start - time.time())
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: elapsed_time is already float

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: elapsed_time is already float

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 11]
```diff
@@ -0,0 +1,368 @@
...
+                Note: the number of epoch is 1 indexed. (Results after the first epoch: epoch = 1)
+
+            Fidelity parameters, check get_fidelity_space(). Uses default (max) value if None.
+        rng : np.random.RandomState, int, None
+            Random seed to use in the benchmark.
+
+            To prevent overfitting on a single seed, it is possible to pass a
+            parameter ``rng`` as 'int' or 'np.random.RandomState' to this function.
+            If this parameter is not given, the default random state is used.
+
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:291]: `shuffle`is missing

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:291]: `shuffle`is missing

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 12]
```diff
@@ -0,0 +1,368 @@
...
+                                rng: Union[np.random.RandomState, int, None] = None,
+                                shuffle: bool = False,
+                                **kwargs) -> Dict:
+        """
+        Get the validated results. Runs a given configuration on the largest budget (here: 50).
+        Parameters
+        ----------
+        configuration
+        fidelity: Dict, None
+            epoch: int - Values: [1, 81]
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:280]: We only allow 81. 

Maybe this is something we should discuss: 
It could be okay to query also the performance for different epoch steps

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:280]: We only allow 81. 

Maybe this is something we should discuss: 
It could be okay to query also the performance for different epoch steps

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 13]
```diff
@@ -0,0 +1,368 @@
...
+                log_perplexity : float,
+                negative_log_perplexity : float,
+                training_cost : float,
+                valid_cost : float,
+                test_cost : float,
+                fidelity : Dict
+                    used fidelities in this evaluation
+        """
+
+        # The result dict should contain already all necessary information -> Just swap the function value from valid
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: outdated comment.

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: outdated comment.

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 14]
```diff
@@ -0,0 +1,368 @@
...
+        # to test and the corresponding time cost
+        assert fidelity['epoch'] == 81, 'Only test data for the 50. epoch is available. '
+        ts_start = time.time()
+
+        self.rng = rng_helper.get_rng(rng)
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]:
```suggestion
        self.rng = rng_helper.get_rng(self.rng, rng)
```

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]:
```suggestion
        self.rng = rng_helper.get_rng(self.rng, rng)
```

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 15]
```diff
@@ -0,0 +1,368 @@
...
+                         'log_perplexity': log_perplexity,
+                         'perplexity': perplexity,
+                         'negative_log_perplexity': neg_log_perplexity,
+                         'training_cost': train_time,
+                         'test_cost': eval_time,
+                         'fidelity': fidelity
+                         }
+                }
+
+    __all__ = ["LanguageModelBenchmark"]
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:383]: Add a new line at the end of file.

-> checkstyle error

Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py:383]: Add a new line at the end of file.

-> checkstyle error

File: hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+MAINTAINER sharmaa@informatik.uni-freiburg.de
+VERSION v0.0.1
+
+%post
+    apt update -y
+    apt install build-essential git wget -y
+
+    cd /home \
+    && mkdir data && cd data \
+    && wget https://raw.githubusercontent.com/pytorch/examples/master/word_language_model/data/wikitext-2/train.txt \
```
Comment (PhMueller) [inline hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark:14]: no need to download in container. The dataloader should download the data automatically to the users host system (defined data dir in the hpobenchrc file)

Comment (PhMueller) [inline hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark:14]: no need to download in container. The dataloader should download the data automatically to the users host system (defined data dir in the hpobenchrc file)

File: hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+    && wget https://raw.githubusercontent.com/pytorch/examples/master/word_language_model/data/wikitext-2/valid.txt \
+    && wget https://raw.githubusercontent.com/pytorch/examples/master/word_language_model/data/wikitext-2/test.txt \
+    && cd /home \
+    && git clone https://github.com/ayushi-3536/HPOBench.git \
```
Comment (PhMueller) [inline hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark:18]: please link to automl and master

Comment (PhMueller) [inline hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark:18]: please link to automl and master

File: hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+    && cd HPOBench \
+    && git checkout fair_adult \
```
Comment (PhMueller) [inline hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark]: shoudl be `lm`

Comment (PhMueller) [inline hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark]: shoudl be `lm`

File: hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark [hunk 19]
```diff
@@ -0,0 +1,30 @@
...
+    && pip install .[adult] \
+    && cd / \
+    && mkdir /var/lib/hpobench/ \
+    && chmod -R 777 /var/lib/hpobench/ \
+    && rm -rf /var/lib/apt/lists/* \
+    && pip cache purge
+
+
+%runscript
+    python -s /home/HPOBench/hpobench/container/server_abstract_benchmark.py mo.adult_benchmark $@
```
Comment (PhMueller) [inline hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark]: Adult is still linked here.

Comment (PhMueller) [inline hpobench/container/recipes/mo/Singularity.LanguageModelBenchmark]: Adult is still linked here.

File: hpobench/util/data_manager.py [hunk 20]
```diff
@@ -849,6 +848,77 @@ def _load(self) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndar
...
+        y_val: np.ndarray
+        X_test: np.ndarray
+        y_test: np.ndarray
+        """
+
+        t = time()
+        self._download()
+        self.X_train, self.X_valid, self.X_test = self._load()
+        self.logger.info(f'LanguageModelDataManager: Data successfully loaded after {time() - t:.2f}')
+        print(self.X_train.shape)
```
Comment (PhMueller) [inline hpobench/util/data_manager.py]: Please remove print statements

Comment (PhMueller) [inline hpobench/util/data_manager.py]: Please remove print statements

**Review by PhMueller — CHANGES_REQUESTED** [review]
I left some commments. could you please have a look at them. Thanks

File: extra_requirements/lm_benchmark.json [hunk 21]
```diff
@@ -0,0 +1,6 @@
+{
+  "lm": [
+    "torch==1.3.0",
+    "tqdm"
```
Comment (PhMueller) [inline extra_requirements/lm_benchmark.json]: add a reasonable version number (you can use the same as for the flower cnn benchmark)

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 22]
```diff
@@ -0,0 +1,383 @@
...
+        _, test_acc = model.eval_fun(self.ntokens, criterion, test_data)
+        eval_test_runtime = time.time() - start_time
+
+        perplexity = math.exp(best_val_loss)
+        log_perplexity = best_val_loss
+        neg_log_perplexity = 10 - best_val_loss
+        elapsed_time = ts_start - time.time()
+
+        return {'function_value': {'log_perplexity': log_perplexity,
+                                   'accuracy': val_acc.item(),
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: please change to `1 - val_acc` so that it is a min problem.

File: hpobench/benchmarks/mo/lm_benchmark.py [hunk 23]
```diff
@@ -0,0 +1,383 @@
...
+                # Anneal the learning rate if no improvement has been seen in the validation dataset.
+                learning_rate /= learning_rate_factor
+
+        perplexity = math.exp(best_test_loss)
+        log_perplexity = best_test_loss
+        neg_log_perplexity = 10 - best_test_loss
+        elapsed_time = ts_start - time.time()
+
+        return {'function_value': {'log_perplexity': log_perplexity,
+                                   'accuracy': test_acc.item(),
```
Comment (PhMueller) [inline hpobench/benchmarks/mo/lm_benchmark.py]: see objective func

File: hpobench/container/benchmarks/mo/lm_benchmark.py [hunk 24]
```diff
@@ -0,0 +1,12 @@
+""" Benchmark for the Multi-Objective Language Model Benchmark from hpobench/benchmarks/mo/lm_benchmark.py
+"""
+
+from hpobench.container.client_abstract_benchmark import AbstractBenchmarkClient
+
+
+class LanguageModelBenchmark(AbstractBenchmarkClient):
```
Comment (PhMueller) [inline hpobench/container/benchmarks/mo/lm_benchmark.py]: Use here `AbstractMoBenchmarkClient`

File: hpobench/container/benchmarks/mo/lm_benchmark.py [hunk 25]
```diff
@@ -0,0 +1,12 @@
...
+    def __init__(self, **kwargs):
+        kwargs['benchmark_name'] = kwargs.get('benchmark_name', 'LanguageModelBenchmark')
+        kwargs['container_name'] = kwargs.get('container_name', 'lm_benchmark')
+        kwargs['latest'] = kwargs.get('container_tag', '0.0.1')
```
Comment (PhMueller) [inline hpobench/container/benchmarks/mo/lm_benchmark.py:11]: add

`kwargs['gpu'] = kwargs.get('gpu', True)`

File: hpobench/dependencies/lm/model.py [hunk 26]
```diff
@@ -0,0 +1,148 @@
...
+            >>> output = pos_encoder(x)
+        """
+        x = x + self.pe[:x.size(0), :]
+        return self.dropout(x)
+
+
+class TransformerModel(nn.Module):
+    """Container module with an encoder, a transformer module, and a decoder."""
+
+    def __init__(self, ntoken, ninp, nhead, nhid, nlayers, dropout=0.5, bptt=35, rng=None):
```
Comment (PhMueller) [inline hpobench/dependencies/lm/model.py:51]: signature if possible

**Review by PhMueller — CHANGES_REQUESTED** [review]
it would be cool if you could go through the comments. Thanks
